openssl/src/ssl/mod.rs

```diff
@@ -2944,7 +3059,68 @@ where
         }
     }
 
-    // Future work: early IO methods
+    /// Initiates the handshake.
+    ///
+    /// This will fail if `set_accept_state` or `set_connect_state` was not called first.
+    ///
+    /// This corresponds to [`SSL_do_handshake`].
+    ///
+    /// [`SSL_do_handshake`]: https://www.openssl.org/docs/manmaster/man3/SSL_do_handshake.html
+    pub fn handshake(self) -> Result<SslStream<S>, HandshakeError<S>> {
```

Comment:
Do connect and accept not work if called after set_accept_state or set_connect_state?

Reply:
`SSL_connect` and `SSL_accept` are both thin wrappers around `SSL_do_handshake` that initialize the connection with the corresponding state if it has not previously had a state set. Setting one state and then using the other should be harmless, albeit confusing, although the documentation doesn't guarantee this.

It might be cleaner to do away with the simplified calls for this interface entirely, but I presume we're stuck with them.

Reply:
Ok, seems reasonable to keep this then!

Reply:
If I were to rebuild this API I'd probably make `accept` and `connect` constructors and get rid of `new` and the `SSL_accept`/`connect` wrappers.